junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/EngineDiscoveryOrchestrator.java

```diff
@@ -173,6 +202,15 @@ private void logTestDescriptorExclusionReasons(Map<String, List<TestDescriptor>>
 		});
 	}
 
+	private Map<TestEngine, TestDescriptor> pruneEngines(Map<TestEngine, TestDescriptor> result) {
+		// @formatter:off
+		return result.entrySet()
+				.stream()
+				.filter(entry -> !entry.getValue().getChildren().isEmpty())
```

Comment:
❌ Although I don't know of any engine that is completely dynamic, we should call `containsTests` here:
```suggestion
				.filter(entry -> TestDescriptor.containsTests(entry.getValue())
```